recovery/elliptics_recovery/types/merge.py

```diff
@@ -686,6 +683,130 @@ def succeeded(self):
+            if ctx.backend_id is not None:
+                ranges = group_routes.get_address_backend_ranges(ctx.address, ctx.backend_id)
+                routes = [(ctx.address, ctx.backend_id, sort_ranges(ranges))]
+            else:
+                for backend_id in group_routes.get_address_backends(ctx.address):
+                    ranges = group_routes.get_address_backend_ranges(ctx.address, backend_id)
+                    routes.append((ctx.address, backend_id, sort_ranges(ranges)))
+        else:
+            for addr, backend_id in group_routes.addresses_with_backends():
+                ranges = group_routes.get_address_backend_ranges(addr, backend_id)
+                routes.append((addr, backend_id, sort_ranges(ranges)))
+
+        log.info("Server-send recovery: group: {0}, num addresses: {1}".format(group, len(routes)))
+        return routes
+
+    def recover(self, keys):
+        '''
+        Tries to recover keys from every backend via server-send. Then it
+        removes keys with older timestamp or invalid checksum.
+        Returns list of keys that was not recovered via server-send.
+        '''
+        log.info("Server-send bucket: num keys: {0}".format(len(keys)))
+
+        def contain(key, ranges):
+            index = bisect(ranges, key)
+            return index % 2 == 1
+
+        responses = dict([(str(k), []) for k in keys]) # key -> [list of responses]
+        for addr, backend_id, backend_ranges in self.routes:
+            key_candidates = [k for k in keys if not contain(k, backend_ranges)]
+            if key_candidates:
+                self._server_send(key_candidates, addr, backend_id, responses)
+
+        self._remove_bad_keys(responses)
+        return self._get_unrecovered_keys(responses)
+
+    def _server_send(self, keys, addr, backend_id, responses):
+        '''
+        Calls server-send with a given list of keys to the specific backend.
+        '''
+        log.debug("Server-send: address: {0}, backend: {1}, num keys: {2}".format(addr, backend_id, len(keys)))
+
+        self.session.set_direct_id(addr, backend_id)
+        iterator = self.session.server_send(keys, elliptics.iterator_flags.move, list(self.session.groups))
+        for result in iterator:
+            status = result.response.status
+            key = result.response.key
+            r = (key, status, addr, backend_id)
+            log.debug("Server-send result: key: {0}, status: {1}".format(str(key), status))
+            responses[str(key)].append(r)
+
+    def _remove_bad_keys(self, responses):
+        '''
+        Removes invalid keys with older timestamp or invalid checksum.
+        '''
+        bad_keys = []
+        for val in responses.itervalues():
+            bad_keys.extend([r for r in val if self._check_bad_key(r)])
+
+        results = []
+        for k in bad_keys:
+            key, _, addr, backend_id = k
+            self.session.set_direct_id(addr, backend_id)
+            result = self.session.remove(key)
```

Comment:
Recovery has a safe mode (`-S/--safe`) when it doesn't remove keys. Please check that it isn't set before removing keys and before setting `elliptics.iterator_flags.move` flag.